networkx/algorithms/approximation/steinertree.py

```diff
@@ -113,10 +113,24 @@ def _kou_steiner_tree(G, terminal_nodes, weight):
 
 
 def _remove_nonterminal_leaves(G, terminals):
-    terminals_set = set(terminals)
-    for n in list(G.nodes):
-        if n not in terminals_set and G.degree(n) == 1:
-            G.remove_node(n)
+    terminal_set = set(terminals)
+    leaves = {
+        n
+        for n in G.nodes()
+        if len(list(G.neighbors(n))) - list(G.neighbors(n)).count(n) == 1
+    }
+    nonterminal_leaves = leaves - terminal_set
+
+    while nonterminal_leaves:
+        nbrs = ([nbr for nbr in G.neighbors(n) if n != nbr] for n in nonterminal_leaves)
+        possible_nonterminal_leaves = set().union(*nbrs) - terminal_set
+        G.remove_nodes_from(nonterminal_leaves)
+        leaves = {
+            n
+            for n in possible_nonterminal_leaves
+            if len(list(G.neighbors(n))) - list(G.neighbors(n)).count(n) == 1
+        }
```

Comment:
I agree with the idea of limiting the search for "candidate" non-terminal nodes to those which are neighbors with the current set of non-terminal nodes. On the whole, I think this extra processing is worthwhile since it's relevant for any paths in the graph.

I think this can be "simplified" a bit as well - I don't think it'll make a huge difference in performance (though perhaps slightly better as it cuts down on the number of list->set data structures being created), but might be a bit easier to follow (especially with comments):

```suggestion
        # Limit search for candidate non-terminal nodes to neighbors of current
        # non-terminal nodes
        candidate_leaves = set.union(*(set(G[n]) for n in nonterminal_leaves))
        candidate_leaves -= (nonterminal_leaves | terminal_set)
        # Remove current set of non-terminal nodes
        G.remove_nodes_from(nonterminal_leaves)
        # Find any new non-terminal nodes from the set of candidates
        leaves = {n for n in candidate_leaves if len(set(G[n]) - {n}) == 1}
```